File: src/test/java/org/edtp/universe/region/BlockRegionTest.java
package org.edtp.universe.region;

import net.minecraft.core.BlockPos;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

class BlockRegionTest {
    @Test
    void radiusIsAHalfLengthAndIncludesBothEndpoints() {
        var region = BlockRegion.centered(new BlockPos(10, 64, -10), 2);

        assertEquals(8, region.getMinX());
        assertEquals(12, region.getMaxX());
        assertEquals(62, region.getMinY());
        assertEquals(66, region.getMaxY());
        assertEquals(-12, region.getMinZ());
        assertEquals(-8, region.getMaxZ());
        assertEquals(125L, region.getVolume());
    }

    @Test
    void linearCursorVisitsFirstAndLastBlock() {
        var region = BlockRegion.centered(BlockPos.ZERO, 1);
        var cursor = new BlockPos.MutableBlockPos();

        assertEquals(new BlockPos(-1, -1, -1), region.position(0, cursor).immutable());
        assertEquals(new BlockPos(1, 1, 1), region.position(region.getVolume() - 1, cursor).immutable());
    }
}

File: src/test/kotlin/org/edtp/universe/region/BlockRegionTest.kt
package org.edtp.universe.region

import net.minecraft.core.BlockPos
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test

class BlockRegionTest {
    @Test
    fun `radius is a half length and includes both endpoints`() {
        val region = BlockRegion.centered(BlockPos(10, 64, -10), 2)

        assertEquals(8, region.minX)
        assertEquals(12, region.maxX)
        assertEquals(62, region.minY)
        assertEquals(66, region.maxY)
        assertEquals(-12, region.minZ)
        assertEquals(-8, region.maxZ)
        assertEquals(125L, region.volume)
    }

    @Test
    fun `linear cursor visits first and last block`() {
        val region = BlockRegion.centered(BlockPos.ZERO, 1)
        val cursor = BlockPos.MutableBlockPos()

        assertEquals(BlockPos(-1, -1, -1), region.position(0, cursor).immutable())
        assertEquals(BlockPos(1, 1, 1), region.position(region.volume - 1, cursor).immutable())
    }
}
